### DISCUSSION TITLE
How to run Hyperspace IBC relayer

### DISCUSSION THREAD

**Discussion by Farhad-Shabani** [informalsystems/hermes#2948] [Non-SDK support]
## 1. Outline

This manual prepared to serve as a reference for anyone interested in trying another Rust-based IBC relayer, but in Polkadot ecosystem. Here are instructions for running Hyperspace IBC relayer on your local machine. 

## 2. Context
Hyperspace is an off-chain relayer component of ComposableFi’s [Centauri](https://github.com/ComposableFi/centauri) bridge protocol. It is an implementation of IBC in Rust, aiming to relay packets between any IBC-enabled chains, but, in its current shape, it already supports Parachains.

## 3. System requirements

- At least 16GB RAM → 3 ~ 4 GB RAM for each of running node (5 ~ 7 nodes)
- The more cores allocated, the better, but don't allocate all, it can freeze the device
- At least 70 GB of free disk space

## 4. Prerequisites

1. Install Required packages and Rust (Stable & Nightly)
    
    ```bash
    brew update
    brew install openssl
    brew install protobuf
    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
    rustup update nightly
    rustup target add wasm32-unknown-unknown --toolchain nightly
    ```
    
2. **Turn on Rust backtrace and skip building the** `wasm`
    
    ```bash
    RUST_BACKTRACE=full \
    SKIP_WASM_BUILD=1 \ 
    ```
   
You can also check out [this](https://docs.substrate.io/install/) reference to ensure that you are not missing anything

### 5.1. Using Docker Image

Use the below bash script to create `docker-compose.yml` for running a leaner version of chains including 3 nodes for the relay chain and only one collator per Parachain”

```bash
git clone https://github.com/ComposableFi/centauri.git
scripts/build-lean-setup.sh
scripts/run-lean-setup.sh
```

For the complete set of nodes (more powerful hardware required!), You can try this:

```bash
git clone https://github.com/ComposableFi/centauri.git	
cd scripts/parachain-launch
docker-compose up -d
```

<aside>
⚠️ Polkadot docker images should be run on systems enabled for X86_64 (AMD64)
So, it is not compatible with macOS with ARM64, even with the virtualization turned on.
As a result, most of the nodes required for running Parachains and Relay chain are dropped upon running
</aside>

</br>

<aside>
⚠️ If docket-compose failed, Check it for possible ports conflicts & duplicity
</aside>

### 5.2. Manual Instruction

The following steps guide you through manually building the necessary binaries and running the nodes:

1. **Build Composable binary**
    
    ```bash
    git clone https://github.com/ComposableFi/composable.git
    git checkout centauri
    make release   # Execute in the root of the composable repo 
    ```
    
2. **Build Parachain binary**
    
    ```bash
     git clone https://github.com/ComposableFi/centauri.git
     cargo build -p parachain-node --release
    ```
    
3.  **Run Script**
    
    ```bash
    cd centauri/scripts/polkadot-launch
    yarn install
    yarn dev
    ```
    

## 6. **Run Hyperspace test suite**
Run the following command to relay test packets between running Parachains through `hyperspace-testsuite` package

```bash
cargo +nightly test -p hyperspace-testsuite --test parachain_parachain -- --nocapture
```

## Highlights

- Binaries are built in a time-consuming manner (1~2 hours), Particularly if you want to run it through the docker image.
- There are several concrete dependencies and patches (especially around `tendermint-rs` and `ibc-rs`) that should be closely watched to avoid any bugs or conflicts.

**Comment by Wizdave97** [comment]
Superb, thank you @Farhad-Shabani 
 For the manual instructions, let them use the parachain node in centauri
 ```
 Build Parachain binary

git clone https://github.com/ComposableFi/centauri.git
cargo build -p parachain-node --release
```

```
Run Script

cd centauri/scripts/polkadot-launch
yarn install
yarn dev
```

**Reply by Farhad-Shabani** [reply]
Thanks for noting that. I updated it
